src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -0,0 +1,255 @@
+            return hashLength;
+        }
+
+        private static void Extract(HashAlgorithmName hashAlgorithmName, int hashLength, ReadOnlySpan<byte> ikm, ReadOnlySpan<byte> salt, Span<byte> prk)
+        {
+            if (prk.Length != hashLength)
+            {
+                throw new ArgumentException(nameof(prk));
+            }
+
+            Debug.Assert(HashLength(hashAlgorithmName) == hashLength);
+
+            using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, salt))
+            {
+                hmac.AppendData(ikm);
+                GetHashAndReset(hmac, prk);
+            }
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Expand function
+        /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="prk">Pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step)</param>
+        /// <param name="outputLength">Length of the output keying material</param>
+        /// <param name="info">Optional context and application specific information</param>
+        /// <returns>Output keying material</returns>
+        public static byte[] Expand(HashAlgorithmName hashAlgorithmName, byte[] prk, int outputLength, byte[] info = null)
+        {
+            if (prk == null)
+                throw new ArgumentNullException(nameof(prk));
+
+            int hashLength = HashLength(hashAlgorithmName);
+
+            if (outputLength <= 0 || outputLength > 255 * hashLength)
+                throw new ArgumentOutOfRangeException(nameof(outputLength));
+
+            info = info ?? Array.Empty<byte>();
+            byte[] result = new byte[outputLength];
+            Expand(hashAlgorithmName, hashLength, prk, result, info);
+
+            return result;
+        }
+
+        /// <summary>
+        /// Performs the HKDF-Expand function
+        /// See section 2.3 of <a href="https://tools.ietf.org/html/rfc5869#section-2.3">RFC5869</a>
+        /// </summary>
+        /// <param name="hashAlgorithmName">Hash algorithm used for HMAC operations</param>
+        /// <param name="prk">Pseudorandom key of at least <see cref="HashLength"/> bytes (usually the output from Expand step)</param>
+        /// <param name="output">Output buffer representing output keying material</param>
+        /// <param name="info">Context and application specific information (can be an empty span)</param>
+        public static void Expand(HashAlgorithmName hashAlgorithmName, ReadOnlySpan<byte> prk, Span<byte> output, ReadOnlySpan<byte> info)
+        {
+            int hashLength = HashLength(hashAlgorithmName);
+
+            // We want to throw different exception type for this overload
+            if (output.Length > 255 * hashLength)
+                throw new ArgumentException(nameof(output));
```

Comment:
ArgumentException's first parameter is message, not argument name.

Reply:
This might be something to add an analyzer for. using nameof to provide as the first parameter of ArgumentException is almost certainly wrong.

Reply:
Right, yes, analyzer would be useful 🤦‍♂

Reply:
Yes, that's one of the examples I have enumerated in https://github.com/dotnet/corefx/issues/40739